### PR TITLE
Ex2

### LINKED ISSUE #5: ex2
A web scrapping excercise in Python.

Dod: working code solution.

### PR DESCRIPTION
**Pull request by yarin177**
- Added functions to retrieve HTML, resorces and screenshots
- Added a function to handle output folder (whenever it needs to be removed after every time the program starts)
- Added a function to create the output directory
- Added a function that writes a JSON dictonary to a file
- Added a function to create a local folder for every url
- Added a function that takes a URL and scarpping the page by using all of the other functions
- Added support for threads pool, each webpage is being scrapped with another thread
- Added a main function to start all threads
- Added local path variables to make it more readable
- Added neccesary comments

### CODE REVIEW

File: ex2-web-scrapping/browser.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+def main():  
+    urls = get_urls_from_file()
+    init_output_folder()
+
+    with ThreadPoolExecutor(max_workers=8) as executor: # init a thread pool executor
+        executor.map(scrape_page, urls) # for each of the urls start func scrape_page
+
+if __name__ == "__main__":
+    main()
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py:76]: One line break after code, use the library: prospector with everything, it checks your code (quality and more)

File: ex2-web-scrapping/browser.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    data["resources"] = get_external_links(browser)
+    data["screenshot"] = browser.get_screenshot_as_base64() 
+    browser.save_screenshot(os.path.join(url_path, 'screenshot.png'))
+    write_json(url_path,data)
+
+def main():  
+    urls = get_urls_from_file()
+    init_output_folder()
+
+    with ThreadPoolExecutor(max_workers=8) as executor: # init a thread pool executor
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Whats the difference between ThreadPool and ThreadPoolExecuter?

File: ex2-web-scrapping/browser.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import os
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py:1]: General Comments:
- use prospector and pydocstyle to check your code
- Write comments with google style on every function

File: ex2-web-scrapping/browser.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import json
+from selenium import webdriver
+from concurrent.futures import ThreadPoolExecutor
+
+INPUT_FOLDER = 'input' 
+INPUT_FILE = os.path.join(INPUT_FOLDER, 'urls.input')
+OUTPUT_FOLDER = 'output'
+
+def delete_output_dir():
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Check if there is a function in os that does recursive delete

File: ex2-web-scrapping/browser.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    for url_folder in os.listdir(OUTPUT_FOLDER): # for every url folder in output
+        local_folder_path = os.path.join(OUTPUT_FOLDER, url_folder)
+        for file in os.listdir(local_folder_path): # delete every file in url dir
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Line break before for/if/etc

File: ex2-web-scrapping/browser.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+            os.remove(os.path.join(local_folder_path, file))
+        os.rmdir(local_folder_path) # eventually remove the empty url dir
+
+def create_url_folder(index: str):
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: There is no need for function for one line

Comment (IdoFriedman1) [reply ex2-web-scrapping/browser.py]: See if you did it again

File: ex2-web-scrapping/browser.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    os.mkdir(os.path.join(OUTPUT_FOLDER, f"url_{str(index+1)}"))
+
+def get_html(browser: webdriver.Chrome) -> str:
+    return browser.page_source
+
+def get_external_links(browser: webdriver.Chrome) -> list:
+    imageLinks = [image.get_attribute("src") for image in browser.find_elements_by_xpath("//a//img")]
+    website_urls = [elem.get_attribute("href") for elem in browser.find_elements_by_xpath("//a[@href]")]
+    return imageLinks + website_urls
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Lone break before return, and maybe use extend?

File: ex2-web-scrapping/browser.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+def init_output_folder():
+    if not os.path.isdir(OUTPUT_FOLDER): # if output dir doesn't exists
+        os.mkdir(OUTPUT_FOLDER)
+    else:
+        delete_output_dir() # delete everything inside
+
+def get_urls_from_file() -> list:
+    urls = open(INPUT_FILE,'r').read().split('\n')
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: You didn’t close the file

File: ex2-web-scrapping/browser.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    urls_index = [i for i in range(len(urls))]
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Range gives you that list…

File: ex2-web-scrapping/browser.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    return list(zip(urls, urls_index))
+
+def write_json(url_path: str, json_data: dict):
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: You don’t need a function for this small action

File: ex2-web-scrapping/browser.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    with open(os.path.join(url_path, 'browse.json'), 'w',encoding='utf-8') as f:
+        json.dump(json_data, f)
+
+def scrape_page(url: list):
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py:45]: Comments and spaces by logic

Comment (IdoFriedman1) [reply ex2-web-scrapping/browser.py:45]: Comments above row

File: ex2-web-scrapping/browser.py (hunk 2 above)
Comment (yarin177) [reply ex2-web-scrapping/browser.py]: A thread pool is a software design pattern for achieving concurrency of execution in a computer program.
ThreadPoolExecutor is a Python implementation of ThreadPool, the Executor subclass uses a pool of threads to execute calls asynchronously.

File: ex2-web-scrapping/browser.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    Returns:
+      list: A list of strings contaning external links and images
+    """
+    imageLinks = [image.get_attribute("src") for image in browser.find_elements_by_xpath("//a//img")]
+    website_urls = [elem.get_attribute("href") for elem in browser.find_elements_by_xpath("//a[@href]")]
+    
+    return imageLinks.extend(website_urls)
+
+def init_output_folder():
+
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Unnecessary line break

File: ex2-web-scrapping/browser.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+from selenium import webdriver
+from concurrent.futures import ThreadPoolExecutor
+from shutil import rmtree
+
+INPUT_FOLDER = 'input' 
+INPUT_FILE = os.path.join(INPUT_FOLDER, 'urls.input')
+OUTPUT_FOLDER = 'output'
+
+def get_external_links(browser: webdriver.Chrome) -> list:
+    """Returns a list of URLs containing all the external links and images in a page'
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Start the docs in the next line not on the “””

File: ex2-web-scrapping/browser.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    
+    return list(zip(urls, urls_index))
+
+def scrape_page(url: list):
+    """Scrapes a page and saves its info in a unique folder inside the 'output' dir
+
+    Args:
+      list: (url) that contains the URL of the page and its index
+
+    Returns:
```
Comment (IdoFriedman1) [inline ex2-web-scrapping/browser.py]: Incase of no return you can remove the return

**Pull request merged**